code/modules/hydroponics/gene_modder.dm

```diff
@@ -107,258 +106,135 @@
-				for(var/a in trait_genes)
-					var/datum/plant_gene/G = a
-					dat += "<tr><td width='260px'>[G.get_name()]</td><td>"
-					if(can_extract)
-						dat += "<a href='?src=[UID()];gene=[G.UID()];op=extract'>Extract</a>"
-					dat += "<a href='?src=[UID()];gene=[G.UID()];op=remove'>Remove</a>"
-					dat += "</td></tr>"
-				dat += "</table>"
-			else
-				dat += "No trait-related genes detected in sample.<br>"
-			if(can_insert && istype(disk.gene, /datum/plant_gene/trait))
-				dat += "<a href='?src=[UID()];op=insert'>Insert: [disk.gene.get_name()]</a>"
-			dat += "</div>"
-
-		dat += "<div class='line'><h3>Variant</h3></div><div class='statusDisplay'><table>"
-		dat += "<tr><td width='260px'>[seed.variant ? seed.variant : "None"]</td>"
-		dat += "<td><a href='?src=[UID()];set_v=1'>Edit</a></td>"
-		if(seed.variant)
-			dat += "<td><a href='?src=[UID()];del_v=1'>Remove</a></td>"
-		dat += "</tr></table></div>"
-	else
-		dat += "<br>No sample found.<br><span class='highlight'>Please, insert a plant sample to use this device.</span>"
-	popup.set_content(dat)
-	popup.open()
+	data["disk"] = list()
+
+	if(disk)
+		var/name = "Empty Disk"
+		if(disk.gene)
+			name = disk.gene.get_name()
+		if(disk.read_only)
+			name = "[name] (Read Only)"
+		data["disk"] = list(
+			"name" = name,
+			"can_insert" = disk.gene?.can_add(seed),
+			"can_extract" = !disk.read_only,
+			"is_core" = istype(disk?.gene, /datum/plant_gene/core)
+		)
+
+	data["modal"] = ui_modal_data(src)
 
+	return data
 
-/obj/machinery/plantgenes/Topic(href, list/href_list)
+/obj/machinery/plantgenes/ui_act(action, list/params, datum/tgui/ui, datum/ui_state/state)
 	if(..())
-		return 1
-	usr.set_machine(src)
-
-	if(href_list["eject_seed"] && !operation)
-		if(seed)
-			seed.forceMove(loc)
-			seed.verb_pickup()
-			seed = null
-			update_genes()
-			update_icon(UPDATE_OVERLAYS)
-		else
-			var/obj/item/I = usr.get_active_hand()
-			if(istype(I, /obj/item/seeds))
-				if(!usr.drop_item())
-					return
-				insert_seed(I)
-				to_chat(usr, "<span class='notice'>You add [I] to the machine.</span>")
-		update_icon(UPDATE_OVERLAYS)
-	else if(href_list["eject_disk"] && !operation)
-		if(disk)
-			disk.forceMove(loc)
-			disk.verb_pickup()
-			disk = null
-			update_genes()
-		else
-			var/obj/item/I = usr.get_active_hand()
-			if(istype(I, /obj/item/disk/plantgene))
-				if(!usr.drop_item())
-					return
-				disk = I
-				disk.forceMove(src)
-				to_chat(usr, "<span class='notice'>You add [I] to the machine.</span>")
-	else if(href_list["op"] == "insert" && disk && disk.gene && seed)
-		if(!operation) // Wait for confirmation
-			operation = "insert"
-		else
-			if(!istype(disk.gene, /datum/plant_gene/core) && disk.gene.can_add(seed))
-				seed.genes += disk.gene.Copy()
-				if(istype(disk.gene, /datum/plant_gene/reagent))
-					seed.reagents_from_genes()
-			update_genes()
-			repaint_seed()
-			operation = ""
-			target = null
+		return
+	if(stat & (NOPOWER|BROKEN))
+		return
```

Comment:
```suggestion
```
`ui_status()` will already set the UI state to `status_close` if the machine has nopower or is broken

```dm
/datum/proc/ui_act(action, list/params, datum/tgui/ui, datum/ui_state/state)
	// If UI is not interactive or usr calling Topic is not the UI user, bail.
	if(!ui || ui.status != STATUS_INTERACTIVE)
		return TRUE
```

```dm
/obj/machinery/ui_status(mob/user, datum/ui_state/state)
	if(!interact_offline && (stat & (NOPOWER|BROKEN)))
		return STATUS_CLOSE

	return ..()
```